src/pmdas/bpf/modules/bashreadline.c

```diff
@@ -0,0 +1,394 @@
+    if (fp)
+        pclose(fp);
+    return result;
+}
+
+static int bashreadline_init(dict *cfg, char *module_name)
+{
+    LIBBPF_OPTS(bpf_object_open_opts, open_opts);
+    off_t func_off;
+    int err;
+    char *val;
+
+    if ((val = pmIniFileLookup(cfg, module_name, "process_count")))
+        env.process_count = atoi(val);
+
+    if (libreadline_path) {
+        readline_so_path = libreadline_path;
+    } else if ((readline_so_path = find_readline_so()) == NULL) {
+        pmNotifyErr(LOG_ERR, "failed to find readline");
+        return 1;
+    }
+
+    err = ensure_core_btf(&open_opts);
+    if (err) {
+        pmNotifyErr(LOG_ERR, "failed to fetch necessary BTF for CO-RE: %s", strerror(-err));
+        return err;
+    }
+
+    obj = bashreadline_bpf__open_opts(&open_opts);
+    if (!obj) {
+        pmNotifyErr(LOG_ERR, "failed to open BPF object");
+        return 1;
+    }
+
+    err = bashreadline_bpf__load(obj);
+    if (err) {
+        pmNotifyErr(LOG_ERR, "failed to load BPF object: %d", err);
+        return err;
+    }
+
+    func_off = get_elf_func_offset(readline_so_path, "readline");
+    if (func_off < 0) {
+        pmNotifyErr(LOG_ERR, "cound not find readline in %s", readline_so_path);
+        return func_off;
+    }
+
+    obj->links.printret = bpf_program__attach_uprobe(obj->progs.printret, true,
+            -1, readline_so_path, func_off);
+    if (!obj->links.printret) {
+        err = -errno;
+        pmNotifyErr(LOG_ERR, "failed to attach readline: %d", err);
+        return err;
+    }
+
+    pb = perf_buffer__new(bpf_map__fd(obj->maps.events), PERF_BUFFER_PAGES,
+            handle_event, handle_lost_events, NULL, NULL);
+    if (!pb) {
+        err = -errno;
+        pmNotifyErr(LOG_ERR, "failed to open perf buffer: %d", err);
+        return err;
+    }
+
+    /* internal/external instance ids */
+    fill_instids(env.process_count, &bashreadline_instances);
+    fill_instids(env.process_count, &bashreadline_lost);
+
+    /* Initialize the tail queue. */
+    TAILQ_INIT(&head);
+
+    return err != 0;
+}
+
+static void bashreadline_shutdown()
+{
+    struct tailq_entry *itemp;
+
+    free(bashreadline_instances);
+    free(bashreadline_lost);
+
+    if (readline_so_path)
+        free(readline_so_path);
+    perf_buffer__free(pb);
+    bashreadline_bpf__destroy(obj);
+    /* Free the entire cache queue. */
+    while ((itemp = TAILQ_FIRST(&head))) {
+        TAILQ_REMOVE(&head, itemp, entries);
+        free(itemp);
+    }
+}
+
+static void bashreadline_refresh(unsigned int item)
+{
+    perf_buffer__poll(pb, PERF_POLL_TIMEOUT_MS);
+}
+
+static int bashreadline_fetch_to_atom(unsigned int item, unsigned int inst, pmAtomValue *atom)
+{
+    struct tailq_entry *value;
+
+    if (inst == PM_IN_NULL) {
+        return PM_ERR_INST;
+    }
+
+    if(!get_item(inst, &value))
+        return PMDA_FETCH_NOVALUES;
+
+    /* bpf.bashreadline.comm */
+    if (item == COMM) {
+        atom->cp = value->str_t.str;
```

Comment:
Since str_t.str is a fixed size array of chars, isn't it a null-terminated string?
https://github.com/performancecopilot/pcp/blob/main/vendor/github.com/iovisor/bcc/libbpf-tools/bashreadline.h#L10

Reply:
Good point - I guess it is in this case, I'm not sure about all cases.